### DISCUSSION TITLE
Rigid Tapping G33.1 for Lathe and Milling

### DISCUSSION THREAD

**Discussion by hanifhan** [grblHAL/core#515] [Q&A]
did G33.1 ( Rigid Tapping ) is supported on Milling and Lathe ? or is there any plan or clue to make it possible because i want to try implement rigid tapping on GRBLHal

**Comment by karoria** [comment]
I want to also try rigid tapping on mill. If it is experimental, then also I can try out.

**Comment by terjeio** [comment]
No it is not supported yet - it is on my longer term todo list though. But nice if somebody else will take on the challenge.
I am not sure the current spindle sync code can be used as-is since I do not know how it will behave during deceleration/acceleration of the spindle. Using a stepper as the spindle motor might be easier to get working?

**Reply by hanifhan** [reply]
hi terje  i made the function in motion control.c for the rigid tapping , but right now im confused on getting initial position of the Z position so it will be back and still in synchronized mode , can u give me any suggestion on getting initial position and set the spindle from cw to ccw ?


void mc_rigid(float *target, plan_line_data_t *pl_data)
{
    // Store the initial Z-axis position ??
    float initial_z_position = pl_data->position[Z_AXIS]; // Assuming pl_data contains the current position

    // Execute the given command
    mc_line(target, pl_data);

    // Wait for the command to complete
    protocol_buffer_synchronize();

    //counter clockwise the spindle ??
    pl_data->spindle.hal->set_state(pl_data->spindle.hal, pl_data->spindle.state.ccw, pl_data->spindle.rpm);


    // Optional delay for dwelling 
    delay_sec(1.00f, DelayMode_Dwell);

    // Move back to the initial Z-axis position
    float return_position[N_AXIS];
    memcpy(return_position, target, sizeof(float) * N_AXIS);
    return_position[Z_AXIS] = initial_z_position;

    //

    mc_line(return_position, pl_data);

    // Wait for the return move to complete
    protocol_buffer_synchronize();
}

**Reply by terjeio** [reply]
Pass the original position to the new function in the same way as it is done for `mc_thread()`.
Spindle is reversed by inverting the ccw flag:
```
pl_data->spindle.state.ccw = !pl_data->spindle.state.ccw;
pl_data->spindle.hal->set_state(pl_data->spindle.hal, pl_data->spindle.state, pl_data->spindle.rpm);
```

**Reply by hanifhan** [reply]
hi terje , do u have any clue to make the spindle synchronization always count the encoder and convert it into feedrate  ,  without waiting for the target programmed spindle ,  im trying to use that kind of thing for rigid tapping

**Reply by hanifhan** [reply]
or maybe save the previous synchronization feedrate and then send it again with paralel of back to origin Z and CCW spindle , do u think that will be do able for rigid tapping ? here is my code
in gcode.c

        case MotionMode_RigidTapping:
        {
            protocol_buffer_synchronize(); // Wait until any previous moves are finished.

            gc_override_flags_t overrides = sys.override.control; // Save current override disable status.

            status_code_t status = init_sync_motion(&plan_data, gc_block.values.k);
            if (status != Status_OK)
                FAIL(status);

            plan_data.spindle.state.synchronized = On;

            mc_line(gc_block.values.xyz, &plan_data);
            protocol_buffer_synchronize();
            plan_data.spindle.state.synchronized = Off;
            mc_rigid(gc_block.values.xyz,&plan_data,gc_state.position);

            protocol_buffer_synchronize();    // Wait until synchronized move is finished,
            sys.override.control = overrides; // then restore previous override disable status.
           // plan_data.spindle.state.synchronized = Off;
        }
        break;

in motion control.c

void mc_rigid(float *target, plan_line_data_t *pl_data, float *position) {
    // Store the original Z position
    float original_z = position[Z_AXIS];

    // Counter clockwise the spindle 
    pl_data->spindle.state.ccw = !pl_data->spindle.state.ccw;
    pl_data->spindle.hal->set_state(pl_data->spindle.hal, pl_data->spindle.state, pl_data->spindle.rpm);



    //protocol_buffer_synchronize();
    // Move the Z-axis back to the original position
    float new_target[N_AXIS]; 
    memcpy(new_target, target, sizeof(float) * N_AXIS); 
    new_target[Z_AXIS] = original_z; 
    if (!mc_line(new_target, pl_data)) {
        return;
    }
   


    protocol_buffer_synchronize();
    pl_data->spindle.state.ccw = !pl_data->spindle.state.ccw;
    pl_data->spindle.hal->set_state(pl_data->spindle.hal, pl_data->spindle.state, pl_data->spindle.rpm);
   // pl_data->spindle.state.synchronized = Off;

**Reply by terjeio** [reply]
> do u have any clue to make the spindle synchronization always count the encoder and convert it into feedrate , without waiting for the target programmed spindle

I think you should wait for the spindle to reach the target RPM before starting. The tricky part is how to handle the direction change.
Switching spindle direction instantly is a no-no? So you'll have to decelerate/accelerate the spindle in a controlled way?

To get any clues about how to do this I would check out how it is done in [LinuxCNC](https://github.com/LinuxCNC/linuxcnc).

**Reply by hanifhan** [reply]
if wait until the spindle reach target RPM after the direction is change ( assume it is inside the tap ) then the tap is going to break , i think it should have been calculate the speed(feedrate) based on encoder speed and always synchronize it? , gonna find some data on linux CNC first
